### DISCUSSION TITLE
Persistent pyright server for incremental checks?

### DISCUSSION THREAD

**Discussion by zhiwei-arbo** [microsoft/pyright#5974] [General]
We are using Pyright for typechecking our monorepo. As our repo increases in size, I am trying to improve the runtimes of running `pyright` on our whole codebase.

One idea was to utilize some Bazel magic to run `pyright` only on files that have changed, which essentially boils down to invoking `pyright` N times for N changed files. This turned out to be really slow - each `pyright` invocation had non-negligible startup time, presumably because it has to start a server analyzing the imported libraries etc.

Is it possible at all to improve this? I don't have enough context on how the Pyright internals work, but I was hoping for something among the lines of having a persistent Pyright server, and hopefully that makes the evaluation of the N files much faster.

**Comment by erictraut** [comment]
The command-line version of pyright is already able to run in a persistent manner. Just run it with the `--watch` command. It will remain running, and it will re-analyze files as they are modified. It retains a graph of dependencies between files, so when a file is modified, it can determine which files are potentially affected and limit its reanalysis to this subset.

Invoking pyright separately for each file will be very slow because every file depends on other files (at the very least, on `builtins.pyi`), so any symbols referenced in the file will require pyright to evaluate those symbols in other files, recursively. When pyright analyzes multiple files, it's able to take advantage of an in-memory cache to avoid re-evaluating the types of symbols that are used in common.

**Reply by zhiwei-arbo** [reply]
Thank you! The `--watch` command does sound like it achieves this partway. Since i'm integrating this as a CI check, I'm looking for a way to start the server, then evaluate the individual files against it, where the files are already modified - I just need to know the result of the check.

Would this already be possible? And if not, can you help with some pointers on how this can be achieved?
